src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java

```diff
@@ -109,9 +197,33 @@ public Step newInstance(Map<String,Object> arguments) throws Exception {
      * @param step a fully-configured step (assignable to {@link #clazz})
      * @return arguments that could be passed to {@link #newInstance} to create a similar step instance
      * @throws UnsupportedOperationException if this descriptor lacks the ability to do such a calculation
+     * @deprecated
+     *      Use {@link #uninstantiate(Step)}
      */
     public Map<String,Object> defineArguments(Step step) throws UnsupportedOperationException {
-        return DescribableModel.uninstantiate_(step);
+        if (Util.isOverridden(StepDescriptor.class, getClass(), "uninstantiate", Step.class)) {
+            // if the subtype has defined the uninstantiate() method, delegate to that
+            return uninstantiate(step).toMap();
+        } else {
+            // otherwise assume legacy usage
+            return DescribableModel.uninstantiate_(step);
+        }
+    }
+
+    /**
+     * Determine which arguments went into the configuration of a step configured through a form submission.
+     * @param step a fully-configured step (assignable to {@link #clazz})
+     * @return arguments that could be passed to {@link #newInstance} to create a similar step instance
+     * @throws UnsupportedOperationException if this descriptor lacks the ability to do such a calculation
+     */
+    public UninstantiatedDescribable uninstantiate(Step step) throws UnsupportedOperationException {
+        if (Util.isOverridden(StepDescriptor.class, getClass(), "uninstantiate", Step.class)) {
+            // newer clients are called older implementations
+            return new UninstantiatedDescribable(defineArguments(step));
```

Comment:
We can, but doing so doesn't buy us much, so we don't need to.

a literal map is a valid part of the object graph in `UninstantiatedDescribable`, and we'll just end up producing that with a map entry named "$class". We cannot reliably look at such a map and recreate a `UninstantiatedDescribable` without knowing the expected type.

We talked about this earlier and we said it's OK for legacy step implementations to behave poorly wrt snippetizer, the primary client of this functionality.